[CL-1196] Flexible pages e2e/unit tests

lilkraftwerk opened the pull request:
Cypress e2e tests and unit tests for the new Flexible Pages feature. Most of the changed files are just adding `data-testid` tags to components, which are only used in locating them for testing purposes and don't affect any other functionality. 

https://user-images.githubusercontent.com/3614128/181176874-43bec5a2-d63e-4809-a14c-c604a66d6155.mov

https://user-images.githubusercontent.com/3614128/181176909-a3999bb4-c91c-4f9e-aa35-71f83686d773.mov

lilkraftwerk committed:
Basic landing page test

lilkraftwerk committed:
Add e2e test for top and bottom info sections

Commit:
Translations updated by CI (extract-intl)

lilkraftwerk committed:
Merge branch 'CL-5-flexible-pages-feature-FE' of github.com:CitizenLabDotCo/citizenlab into CL-1196-flexible-pages-e-2-e-tests

lilkraftwerk committed:
Merge branch 'CL-1196-flexible-pages-e-2-e-tests' of github.com:CitizenLabDotCo/citizenlab into CL-1196-flexible-pages-e-2-e-tests

lilkraftwerk committed:
Add e2e test for top and bottom info sections

lilkraftwerk committed:
Revert translations file to parent branch

lilkraftwerk committed:
Add test for event widget display

lilkraftwerk committed:
Add e2e test for hero banner settings

lilkraftwerk committed:
Finalize e2e tests for flexible pages

lilkraftwerk committed:
Merge branch 'CL-1195-flexible-pages-unit-tests' into CL-1196-flexible-pages-e-2-e-tests

lilkraftwerk committed:
Check color and opacity of hero banner

lilkraftwerk committed:
Format with Prettier

lilkraftwerk committed:
Remove topinfosection unit test

lilkraftwerk commented on front/cypress/support/commands.ts:
```diff
@@ -1042,7 +1043,7 @@ export function apiCreateEvent({
           },
           start_at: startDate.toJSON(),
-          end_at: startDate.toJSON(),
+          end_at: endDate.toJSON(),
```
This was a typo in the existing codebase, my linter pointed it out

lilkraftwerk committed:
Use homepage settings instead of appConfig

lilkraftwerk committed:
Merge branch 'CL-5-flexible-pages-feature-FE' of github.com:CitizenLabDotCo/citizenlab into CL-1196-flexible-pages-e-2-e-tests

IvaKop commented on front/app/components/ContentContainer/index.tsx:
```diff
@@ -39,13 +39,14 @@ interface Props {
 export default class ContentContainer extends PureComponent<Props, State> {
   render() {
-    const { mode, id, className, children } = this.props;
+    const { mode, id, className, children, testId } = this.props;
```
Passing data-testid as a prop feels a bit out of place - it rarely needs to be dynamic. Also - normally you would get and test what is inside the container - not the container itself?

IvaKop commented on front/app/components/admin/SubmitWrapper/index.tsx:
```diff
@@ -92,6 +92,7 @@ interface Props
   onClick?: (event: FormEvent<any>) => void;
   buttonStyle?: ButtonStyles;
   animate?: boolean;
+  testId?: string;
```
Here also - there's probably a better way to target this - getByRole('button') perhaps?

IvaKop commented on front/app/containers/Admin/pages-menu/SectionToggle/AdminEditButton.tsx:
```diff
@@ -7,13 +7,19 @@ import messages from './messages';
 
 interface Props {
   onClick: () => void;
+  testId?: string;
```
Same as above comment - get it in another way? `getBytestId` should normally be last resort anyway, although admittedly we are somewhat overusing it currently in other tests, as well.
https://testing-library.com/docs/queries/about/#priority

IvaKop commented on front/app/containers/Admin/pages-menu/SectionToggle/index.tsx:
```diff
@@ -33,11 +34,16 @@ const SectionToggle = ({
     <Row isLastItem={isLastItem}>
       <Box display="flex" alignItems="center" justifyContent="center">
-        <Box visibility={hideToggle ? 'hidden' : 'visible'} mr="20px">
+        <Box
```
I wonder if it would be appropriate to use the new accordion / list item components here?

lilkraftwerk replied:
![Screen Shot 2022-07-27 at 17 48 24](https://user-images.githubusercontent.com/3614128/181293266-103f8fe5-fe17-4bf0-854f-0ad5e1915240.png)

The use case is to hide one of the toggles here, from looking at Storybook I don't think the accordion/list item would work

IvaKop commented on front/app/containers/Admin/pages-menu/containers/HeroBanner/index.tsx:
```diff
@@ -457,7 +461,10 @@ const HeroBannerForm = ({ intl: { formatMessage } }: InjectedIntlProps) => {
-        <SectionField key="avatars">
+        <SectionField
+          key="avatars"
+          data-testid="e2e-banner-avatar-toggle-section"
```
For e2e tests, it's standard to use `data-cy` although the library doesn't strictly require it (unlike rtl). But it creates some confusion when e2e prefix is used with `data-testid` which is associated with unit tests normally. Perhaps we can change it?

IvaKop commented on front/app/containers/LandingPage/BannerButton.tsx:
```diff
@@ -17,6 +17,7 @@ const BannerButton = ({ buttonStyle, ...props }: Props) => (
     buttonStyle={buttonStyle || 'primary-inverse'}
     fontWeight="500"
     padding="13px 22px"
+    data-testid="e2e-cta-banner-button"
```
Same here - change to `data-cy`?

IvaKop commented on front/app/containers/LandingPage/HomepageInfoSection.tsx:
```diff
@@ -46,15 +46,20 @@ type Props = {
   fragmentName:
     | 'pages/homepage_info/content'
     | 'pages/homepage_info/top-content';
+  testId?: string;
```
Could we hard-code the testid, instead of using a prop?

IvaKop commented on front/app/containers/LandingPage/LandingPage.test.tsx:
```diff
@@ -0,0 +1,38 @@
+});
+
+describe('<LandingPage />', () => {
+  it('renders with HomepageSettings for logged out users', () => {
```
Should this include a bit more tests - like are there different scenarios for the different combinations of the homepage settings? Should we check it renders what we expect in the different scenarios?

lilkraftwerk replied:
I ended up covering all of the test cases I thought were relevant with the e2e ones, do you think it makes sense to also cover them here?

IvaKop replied:
Perhaps not, if it's the exact same tests. But normally unit tests would be best suited for the components themselves and e2e tests are for the entire user flow as a whole. Do we have some new/ changed components that handle some more complex / bug prone logic and could benefit from a unit test?

IvaKop commented on front/app/containers/LandingPage/SignedOutHeader/FullWidthBannerLayout.tsx:
```diff
@@ -86,11 +86,15 @@ const FullWidthBannerLayout = ({ className }: Props) => {
-      <Container className={`e2e-signed-out-header ${className}`}>
+      <Container
+        data-testid="e2e-full-width-banner-layout-container"
+        className={`e2e-signed-out-header ${className}`}
```
Could we have one of those two - either className with the e2e prefix (not preffered) or `data-cy` id?

IvaKop commented on front/cypress/e2e/admin/pages_and_menu/update_hero_banner_content.cy.ts:
```diff
@@ -0,0 +1,171 @@
+      '0.55'
+    );
+
+    // log in as admin and reload page
```
Should this be in a separate it block?

IvaKop commented on front/cypress/e2e/admin/pages_and_menu/update_homepage_content.cy.ts:
```diff
@@ -0,0 +1,71 @@
+    cy.setAdminLoginCookie();
+    cy.visit('/admin/pages-menu/');
+    cy.acceptCookies();
+    cy.wait(1000);
```
No arbitrary waits should be needed in tests - please remove. If you want to wait for a specific api call to complete - use aliases to intercept / wait for it: https://docs.cypress.io/guides/core-concepts/variables-and-aliases#Intercepts

IvaKop commented on front/cypress/e2e/admin/pages_and_menu/update_homepage_content.cy.ts:
```diff
@@ -0,0 +1,71 @@
+
+    // toggle top and bottom info sections
+    cy.get('[data-testid="top_info_section_enabled-toggle"]').click();
+    cy.wait(1000);
```
Shoudn't be needed :)

lilkraftwerk replied:
I cannot for the life of me get this section of code (toggling 3 toggles) to work without the `wait(1000)` condition. I have added the two request intercepts, and it seems to `wait` for them successfully (it shows up in the logs), but the tests do not pass without the `wait(1000)` at all. Maybe we can look at this together tomorrow as I'm stumped 😕 

edit: fixed, see below

IvaKop reviewed (changes requested):
Nice work 👍 Two blocking items from my perspective:
- test id-s used in Cypress tests should be passed via `data-cy` and the ones used in RTL tests should be passed via `data-testid`. Otherwise it's difficult to know what is used where and it gets a bit tangled
- there shouldn't be the need to use `wait` in the e2e tests

I also had some other comments / questions but I don't consider them blocking.

lilkraftwerk committed:
Fix update homepage content test

lilkraftwerk committed:
Fix hero banner test

lilkraftwerk committed:
Change testid to data-cy

lilkraftwerk committed:
Change testid to data-cy

lilkraftwerk commented on front/app/components/UI/Button/index.tsx:
```diff
@@ -16,6 +16,7 @@ interface Props extends ButtonProps {
 
 interface ButtonContainerProps extends ComponentLibraryButtonContainerProps {
   'data-testid'?: string;
+  'data-cy'?: string;
```
Given that our codebase is almost entirely made up of generated classnames, I can't really think of a better way to do it than give specific, low-level components (like a button) an explicit ID for testing purposes

lilkraftwerk commented:
Ok, changed all the new `data-testid` to `data-cy` and updated the tests a bit. There's one that I can't get to work without `wait` (see comments) but the rest seem good 👍 

Also, do we have the conventions for testing written down anywhere? I can't find anything about the difference between `testid` and `data-cy` and looking through the tests it seems like we're using normal HTML classes and IDs quite a bit as well.

lilkraftwerk committed:
Wait for class to exist instead of waiting by milliseconds

lilkraftwerk committed:
Merge branch 'CL-5-flexible-pages-feature-FE' of github.com:CitizenLabDotCo/citizenlab into CL-1196-flexible-pages-e-2-e-tests

lilkraftwerk commented on front/cypress/e2e/admin/pages_and_menu/update_homepage_content.cy.ts:
```diff
@@ -0,0 +1,94 @@
+    cy.wait('@getHomePage');
+
+    // wait for the next toggle to become disabled, then enabled again once it finishes toggling
+    cy.get('[data-cy="e2e-admin-section-toggle-bottom_info_section_enabled"]')
```
This is kind of a hack to bypass the `wait(1000)` I had earlier, to wait for the toggle to become disabled (request in progress) and then enabled again (request done). I don't think just waiting for the request works because the toggle has animation on it, so it will still be disabled once the request has finished successfully. Not the prettiest thing but it seems to work.

IvaKop replied:
It's a bit strange - perhaps we could think of a better loading state than disabling the toggles in the future 🤔 but it's ok if this is what you settled on for now.

lilkraftwerk replied:
The initial idea was to avoid weird race conditions or needing complex state management locally, since the request to enable/disable is sent immediately after the user presses the toggle, and all three toggles use the same endpoint/response. Disabling the toggles means they're only ever enabled and "toggled" according to the actual state of the backend response, so I figured the accurate feedback was worth it. If we have a better pattern for this lmk :)

IvaKop replied:
We have a ticket to add some sort of a loading indicator / state to the toggle for these kinds of situations - maybe you can sync with Ben on it and implement it as a future improvement? What we have now works but is not the best UX..
https://citizenlab.atlassian.net/browse/TEC-63

lilkraftwerk committed:
Fix admin nav test

lilkraftwerk committed:
Use correct insights URL

IvaKop reviewed (approved):
Approving this as I see nothing blocking anymore 👍 Perhaps we should add still some more unit tests for specific components but don't have enough overview to give specific suggestions.
